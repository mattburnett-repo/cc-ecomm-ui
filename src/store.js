import { configureStore } from '@reduxjs/toolkit';

import authReducer from './features/auth/authSlice';
import productCategoryReducer from './features/productCategory/productCategorySlice'
import productsListingReducer from './features/products/productsSlice'
import ordersReducer from './features/orders/ordersSlice'
import cartsReducer from './features/carts/cartsSlice'

function rootReducer() {
    return {
        reducer: {
            auth: authReducer,
            productCategories: productCategoryReducer,
            orders: ordersReducer,
            carts: cartsReducer,
            products: productsListingReducer,
        }    
    }
}

const store = configureStore(rootReducer());

export default store;