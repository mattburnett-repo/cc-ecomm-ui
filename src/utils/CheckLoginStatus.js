
import { useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'

import { selectIsAuthenticated } from '../features/auth/authSlice'

export default function CheckLoginStatus () {
    // eslint-disable-next-line
    const isLoggedIn = useSelector(selectIsAuthenticated);

    const history = useHistory()

    if(!isLoggedIn) {
        history.push('/login')

        return false
    } else {
        return true;
    }
}        
