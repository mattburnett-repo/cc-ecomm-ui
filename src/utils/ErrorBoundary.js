import React from 'react'

export default class ErrorBoundary extends React.Component {
    constructor(props) {
      super(props);
      this.state = { hasError: false };
    }
  
    static getDerivedStateFromError(error) {
      // Update state to trigger fallback UI
      return { hasError: true };
    }
  
    componentDidCatch(error, errorInfo) {
        // Log error or send logging data to log management tool
        //   logErrorToMyService(error, errorInfo);
        console.log('ErrorBoundary ', errorInfo)
    }
  
    render() {
      if (this.state.hasError) {
        // Render fallback UI
        return <h1>There was an error. Sorry about that; we're working on it.</h1>;
      }
  
      return this.props.children; 
    }
  }

