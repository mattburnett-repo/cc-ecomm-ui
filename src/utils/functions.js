
// USAGE: formatAsCurrency.format(3.005); // "3.01"
const formatAsCurrency = new Intl.NumberFormat('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
});

const formatDate =   new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "long",
    day: "2-digit"
});

module.exports = {
    formatAsCurrency: formatAsCurrency,
    formatDate: formatDate
}