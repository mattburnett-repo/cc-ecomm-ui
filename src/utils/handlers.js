

// handler function declarations

// function handleAddToCart(e){ 
//     alert('handleAddToCart inside utils product_id: ' + e.target.id + ' quantity: ' + e.quantity)
//     // {() => dispatch(addItemToCurrentCart(id))}
//     // can't useDispatch here.
// } 

function handleCheckout(e){ 
    alert('handleCheckout inside utils ')
} 

function handleSearch(e){
    e.preventDefault()
    // let searchTerms = e.target.value
    alert('handleSearch for search term/s ' + e.target.searchTerms.value + '. \nneed to figure out how to get dispatch() to work from here')
}

function handleOrderClick(e) {
    e.preventDefault()

    let order_id = e.target.order_id.value
    alert('go to order detail component ' + order_id)
}

function handleHomeClick() {
    console.log('handleHomeClick')
}
function handleShowCartClick() {
    console.log('handleShowCartClick')
}
function handleCheckoutClick() {
    console.log('handleCheckoutClick')
}
function handleCancelClick() {
    console.log('handleCanceClick')
}


module.exports = {
    handleHomeClick: handleHomeClick,
    handleShowCartClick: handleShowCartClick,
    handleCheckoutClick: handleCheckoutClick,
    handleCancelClick: handleCancelClick,
    // handleAddToCart: handleAddToCart,
    handleCheckout: handleCheckout,
    handleSearch: handleSearch,
    handleOrderClick: handleOrderClick
} // end module.exports