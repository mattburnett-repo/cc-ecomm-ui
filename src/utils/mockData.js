// https://css-tricks.com/writing-tests-for-react-applications-using-jest-and-enzyme/

module.exports = {
    // someday I will figure out how to do this correctly...
    // getUserData: jest.fn(() => {
    //   return Promise.resolve({
    //   data: [
    //     {
    //       id: 1,
    //       username: 'asdf',
    //       email: 'asdf@asdf.com',
    //       password: 'asdf'
    //     }
    //   ]
    //   })
    // }),
    mockCartsData:[
        {
          "cart": {
              "cart_id": 1,
              "user_id": 1,
              "cart_items": [
                  {
                      "product_id": 1,
                      "product_name": "hedgehog",
                      "product_price": 1,
                      "product_quantity": 1,
                      "product_total_price": 1
                  },
                  {
                      "product_id": 2,
                      "product_name": "maine coon cat",
                      "product_price": 2,
                      "product_quantity": 2,
                      "product_total_price": 4
                  },
                  {
                      "product_id": 3,
                      "product_name": "bulldog",
                      "product_price": 3,
                      "product_quantity": 3,
                      "product_total_price": 9
                  },
                  {
                      "product_id": 4,
                      "product_name": "bunny",
                      "product_price": 4,
                      "product_quantity": 4,
                      "product_total_price": 16
                  }
              ],
              "cart_total_price": {
                  "sum": 30,
                  "cart_id": 1
              }
          }
      }
    ],
    mockCartData: {
      cart: {
        cart_id: 1,
        user_id: 1,
        cart_items: [
          {
            product_id: 1,
            product_name: 'hedgehog',
            product_price: 1,
            product_quantity: 1,
            product_total_price: 1
          },
          {
            product_id: 2,
            product_name: 'maine coon cat',
            product_price: 2,
            product_quantity: 2,
            product_total_price: 4
          },
          {
            product_id: 3,
            product_name: 'bulldog',
            product_price: 3,
            product_quantity: 3,
            product_total_price: 9
          },
          {
            product_id: 4,
            product_name: 'bunny',
            product_price: 4,
            product_quantity: 4,
            product_total_price: 16
          }
      ],
      cart_total_price: { sum: 30, cart_id: 1 }
    }},
    mockOrderData: [{
      "order": {
          "cart": {
              "cart_id": 1,
              "cart_items": [
                  {
                      "product_id": 1,
                      "product_name": "product_name_01",
                      "product_price": 1,
                      "product_quantity": 1,
                      "product_total_price": 1
                  },
                  {
                      "product_id": 2,
                      "product_name": "product_name_02",
                      "product_price": 2,
                      "product_quantity": 2,
                      "product_total_price": 4
                  },
                  {
                      "product_id": 3,
                      "product_name": "product_name_03",
                      "product_price": 3,
                      "product_quantity": 3,
                      "product_total_price": 9
                  },
                  {
                      "product_id": 4,
                      "product_name": "product_name_04",
                      "product_price": 4,
                      "product_quantity": 4,
                      "product_total_price": 16
                  }
              ]
          },
          "user_id": 1,
          "order_id": 1,
          "order_date": null,
          "total_price": {
              "sum": 30,
              "cart_id": 1
          }
        }
      }],
    mockOrderConfirmationData:
      {
        "order_id": 1,
        "total_amount": 1234,
        "order_date": '12-Dez-2021',
        "payment_type": 'Deutschmarks',
        "delivery_address": '123 Katzendamm, 90210 Berlin DE'
      },
    mockProductCategories: [
        // {
        //   "category_id": 0,
        //   "description": "Please select a product category"
        // },
        {
          "category_id": 4,
          "description": "Auto"
        },
        {
            "category_id": 3,
            "description": "Health"
        },
        {
            "category_id": 1,
            "description": "Pets"
        },
        {
            "category_id": 2,
            "description": "Tech"
        }
      ],
    mockProduct:[
        {
          "id": 2,
          "category_id": 2,
          "name": "product_name_02",
          "description": "product_desc_02",
          "image_url": "https://cdn.pixabay.com/photo/2015/03/27/13/16/maine-coon-694730_1280.jpg",
          "price": "2.00"
        },      
      ],
    mockProducts: [
        {
            "id": 1,
            "category_id": 1,
            "name": "product_name_01",
            "description": "product_desc_01",
            "image_url": "https://pixabay.com/images/id-71282/",
            "price": "1.00"
        },
        {
            "id": 2,
            "category_id": 2,
            "name": "product_name_02",
            "description": "product_desc_02",
            "image_url": "http://www.image.com/item/2",
            "price": "2.00"
        },
        {
            "id": 3,
            "category_id": 3,
            "name": "product_name_03",
            "description": "product_desc_03",
            "image_url": "http://www.image.com/item/3",
            "price": "3.00"
        },
        {
            "id": 4,
            "category_id": 4,
            "name": "product_name_04",
            "description": "product_desc_04",
            "image_url": "http://www.image.com/item/4",
            "price": "4.00"
        }
    ],
    mockUserData:[
      {
        "id": 4,
        "user_name": "asdf",
        "password": "$2b$10$OsTX1L59rYgsmAnsrJsXoO3RH3RSfLgEF53YBB3723B9eoELtUa.a",
        "email": "asdf@asdf.com",
        "google_id": null,
        "google_display_name": null,
        "google_first_name": null,
        "google_last_name": null,
        "google_image": null
      }
    ]
  } // end module.exports