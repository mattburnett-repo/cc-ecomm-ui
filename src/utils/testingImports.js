// export default function TestingImports() {
//     return (
//         import { render, screen, fireEvent } from '@testing-library/react';
//         import renderer from 'react-test-renderer';
//         import userEvent from '@testing-library/user-event';

//         import { Provider } from 'react-redux';
//         import store from '../../store'
//     )
// }