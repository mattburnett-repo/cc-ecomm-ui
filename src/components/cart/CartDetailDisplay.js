
import { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useHistory } from "react-router-dom"

import { setCurrentCartTotalPrice } from '../../features/carts/currentCartActions'
import { selectCurrentCart } from '../../features/carts/cartsSlice'

import GoToCheckoutButton from '../../widgets/GoToCheckoutButton'
import { formatAsCurrency } from '../../utils/functions'
import CartItemDisplay from "./CartItemDisplay"
import { StyledCartDetailDisplay } from '../styles/CartDetailDisplay.styled'

// TODO: this should only look at current cart
//      if user wants to look at a saved cart, it should be moved to current cart before
//          coming here

export default function CartDetailDisplay (props) {
    const { data } = props
    // const { handleSaveCart } = props.handlers

    const currentCart = useSelector(selectCurrentCart)
    const dispatch = useDispatch()

    const [cartTotalPrice, setCartTotalPrice] = useState(0.00)

    const history = useHistory()

    useEffect(() => {
        let the_prices = []
        currentCart.forEach(item => the_prices.push(parseFloat(item.item_total_price)))
        const reducer = (previousValue, currentValue) => previousValue + currentValue;
        let cart_total = the_prices.reduce(reducer, 0.00)

        setCartTotalPrice(formatAsCurrency.format(cart_total))                   // UI state
        dispatch(setCurrentCartTotalPrice(formatAsCurrency.format(cart_total)))  // redux state
    // eslint-disable-next-line
    }, [currentCart])

    return (
        <>
            {/* <StyledCartDetailDisplay>
                <header>
                    <h4>Cart Detail Display: {data.length} item/s</h4>
                    <button aria-label="cart-detail-go-back-button" onClick={() => history.goBack() } >Go Back</button>
                    <button aria-label="cart-checkout-button" onClick={() => history.push('/shipping-info')}>Checkout</button>                  
                </header>    
            </ StyledCartDetailDisplay> */}
            <div role="presentation" aria-label="cart-items">
                {data.map((item, index) => (
                    <CartItemDisplay item={item} />
                ))} 
            </div>
            <StyledCartDetailDisplay>
                <footer>
                    <div role="presentation" aria-label="cart-total-price">
                        Cart total: {cartTotalPrice}
                    </div>
                    <GoToCheckoutButton />
                </footer>
            </StyledCartDetailDisplay>
        </>
    )
}

