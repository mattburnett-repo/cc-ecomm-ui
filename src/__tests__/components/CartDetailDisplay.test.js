// 20211228: Wireframe https://wireframe.cc/kXiUKu

import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Provider } from 'react-redux'
import store from '../../store'

import CartDetailDisplay from '../../components/cart/CartDetailDisplay';

import { mockCurrentCartData } from '../../utils/mockData'

// TODO: this weird data 'mount point' problem worries me...
// console.log('mockCartData', mockCartData)
// console.log('mockCartData.cart', mockCartData.cart) // this one...

function mockHandlers() {
    return {
        handleCheckOut: jest.fn()
    }
}

describe('<CartDetailDisplay /> component tests', () => {
    beforeEach(() => render(<Provider store={store}><CartDetailDisplay data={mockCurrentCartData} handlers={mockHandlers()} /></Provider>));
    // test.todo('it should not take so long...')
    it('should render a cart detail display component', () => {
        screen.getByRole('presentation', { name: /cart-detail-display/i})
    })

    it('should render cart items', () => {
        screen.getByRole('presentation', { name: /^cart-items$/i})

        let theVal = screen.getAllByRole('presentation', { name: /^cart-item$/i})
        expect(theVal).toHaveLength(4)
    })
    it('should render item quantity elements', () => {
        let theVal = screen.getAllByRole('presentation', { name: /cart-item-quantity/i})
        expect(theVal).toHaveLength(4)
    })
    it('should render item images', () => {
        let theVal = screen.getAllByRole('img', { name: /cart-item-image/i})
        expect(theVal).toHaveLength(4)
    })
    it('should rener item names', () => {
        let theVal = screen.getAllByRole('presentation', { name: /cart-item-name/i})
        expect(theVal).toHaveLength(4)        
    })
    it('should render item prices', () => {
        let theVal = screen.getAllByRole('presentation', { name: /cart-item-price/i})
        expect(theVal).toHaveLength(4)       
    })
    it('should render item total prices', () => {
        let theVal = screen.getAllByRole('presentation', { name: /cart-item-total/i})
        expect(theVal).toHaveLength(4)        
    })

    it('should select a value from an item-quantity-selector element', () => {
        window.alert = () => {}; // https://stackoverflow.com/questions/55088482/jest-not-implemented-window-alert
        
        let theVals = screen.getAllByRole('presentation', {name: /item-quantity-selector/i});
        expect(theVals).toHaveLength(4)

        fireEvent.change(theVals[0], { target: { value: 7 } })
        let options = theVals[0];
        expect(options[0].selected).toBeFalsy();
        expect(options[1].selected).toBeFalsy();
        expect(options[2].selected).toBeFalsy();
        expect(options[3].selected).toBeFalsy();
        expect(options[4].selected).toBeFalsy();
        expect(options[5].selected).toBeFalsy();
        expect(options[6].selected).toBeTruthy();
        expect(options[7].selected).toBeFalsy();
        expect(options[8].selected).toBeFalsy();
        expect(options[9].selected).toBeFalsy();
    });

    test.todo('should render a cart total price')
    // it('should render a cart total price', () => {
    //     screen.getByRole('presentation', { name: /cart-total-price/i})
    // })

    it('should rencder remove-from-cart buttons', () => {
        let theVal = screen.getAllByRole('button', { name: /remove-from-cart-button/i})
        expect(theVal).toHaveLength(4)
    })
    it('should render Go Back buttons', () => {
        let theVal = screen.getAllByRole('button', { name: /cart-detail-go-back-button/i})
        expect(theVal).toHaveLength(2)
    })
    it('should render checkout buttons', () => {
        let theVal = screen.getAllByRole('button', { name: /cart-checkout-button/i})
        expect(theVal).toHaveLength(2)
    })

    it('should render a save button', () => {
        let theVal = screen.getAllByRole('button', { name: /cart-save-button/i})
        expect(theVal).toHaveLength(1)
    })

    it('renders a snapshot', () => {
        const tree = renderer.create(<Provider store={store}><CartDetailDisplay data={mockCurrentCartData} handlers={mockHandlers()} /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
})