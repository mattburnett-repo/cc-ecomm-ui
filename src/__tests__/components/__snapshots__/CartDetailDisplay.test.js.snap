// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartDetailDisplay /> component tests renders a snapshot 1`] = `
<div
  aria-label="cart-detail-display"
  role="presentation"
>
  <h4>
    Cart Detail Display: 
    4
     item/s
  </h4>
  <button
    aria-label="cart-detail-go-back-button"
    onClick={[Function]}
  >
    Go Back
  </button>
  <button
    aria-label="cart-checkout-button"
  >
    Checkout
  </button>
  <div
    aria-label="cart-items"
    role="presentation"
  >
    <div
      aria-label="cart-item"
      role="presentation"
    >
      <img
        aria-label="cart-item-image"
      />
      <div
        aria-label="cart-item-quantity"
        role="presentation"
      >
        <label
          htmlFor="item-quantity-selector"
        >
          Quantity:
        </label>
        <select
          aria-label="item-quantity-selector"
          defaultValue="0"
          id="item-quantity-selector"
          name="itemQuantity"
          onChange={[Function]}
          role="presentation"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <div
        aria-label="cart-item-name"
        role="presentation"
      >
        Description: 
      </div>
      <div
        aria-label="cart-item-price"
        role="presentation"
      >
        Price: 
      </div>
      <div
        aria-label="cart-item-total"
        role="presentation"
      >
        Item total: 
      </div>
      <button
        aria-label="remove-from-cart-button"
      >
        Remove From Cart
      </button>
    </div>
    <div
      aria-label="cart-item"
      role="presentation"
    >
      <img
        aria-label="cart-item-image"
      />
      <div
        aria-label="cart-item-quantity"
        role="presentation"
      >
        <label
          htmlFor="item-quantity-selector"
        >
          Quantity:
        </label>
        <select
          aria-label="item-quantity-selector"
          defaultValue="0"
          id="item-quantity-selector"
          name="itemQuantity"
          onChange={[Function]}
          role="presentation"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <div
        aria-label="cart-item-name"
        role="presentation"
      >
        Description: 
      </div>
      <div
        aria-label="cart-item-price"
        role="presentation"
      >
        Price: 
      </div>
      <div
        aria-label="cart-item-total"
        role="presentation"
      >
        Item total: 
      </div>
      <button
        aria-label="remove-from-cart-button"
      >
        Remove From Cart
      </button>
    </div>
    <div
      aria-label="cart-item"
      role="presentation"
    >
      <img
        aria-label="cart-item-image"
      />
      <div
        aria-label="cart-item-quantity"
        role="presentation"
      >
        <label
          htmlFor="item-quantity-selector"
        >
          Quantity:
        </label>
        <select
          aria-label="item-quantity-selector"
          defaultValue="0"
          id="item-quantity-selector"
          name="itemQuantity"
          onChange={[Function]}
          role="presentation"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <div
        aria-label="cart-item-name"
        role="presentation"
      >
        Description: 
      </div>
      <div
        aria-label="cart-item-price"
        role="presentation"
      >
        Price: 
      </div>
      <div
        aria-label="cart-item-total"
        role="presentation"
      >
        Item total: 
      </div>
      <button
        aria-label="remove-from-cart-button"
      >
        Remove From Cart
      </button>
    </div>
    <div
      aria-label="cart-item"
      role="presentation"
    >
      <img
        aria-label="cart-item-image"
      />
      <div
        aria-label="cart-item-quantity"
        role="presentation"
      >
        <label
          htmlFor="item-quantity-selector"
        >
          Quantity:
        </label>
        <select
          aria-label="item-quantity-selector"
          defaultValue="0"
          id="item-quantity-selector"
          name="itemQuantity"
          onChange={[Function]}
          role="presentation"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <div
        aria-label="cart-item-name"
        role="presentation"
      >
        Description: 
      </div>
      <div
        aria-label="cart-item-price"
        role="presentation"
      >
        Price: 
      </div>
      <div
        aria-label="cart-item-total"
        role="presentation"
      >
        Item total: 
      </div>
      <button
        aria-label="remove-from-cart-button"
      >
        Remove From Cart
      </button>
    </div>
  </div>
  <button
    aria-label="cart-detail-go-back-button"
    onClick={[Function]}
  >
    Go Back
  </button>
  <button
    aria-label="cart-checkout-button"
  >
    Checkout
  </button>
</div>
`;
